apollo-federation/src/merger/merger.rs

```diff
@@ -199,13 +239,677 @@ impl Merger {
     fn post_merge_validations(&mut self) {
         todo!("Implement post-merge validations")
     }
+
+    /// Merge enum type from multiple subgraphs
+    pub(crate) fn merge_enum(
+        &mut self,
+        sources: Sources<&EnumType>,
+        dest: &mut EnumType,
+    ) -> Result<(), FederationError> {
+        let usage = self.enum_usages.get(&dest.name.to_string()).cloned().unwrap_or_else(|| {
+            // If the enum is unused, we have a choice to make. We could skip the enum entirely (after all, exposing an unreferenced type mostly "pollutes" the supergraph API), but
+            // some evidence shows that many a user have such unused enums in federation 1 and having those removed from their API might be surprising. We could merge it as
+            // an "input-only" or as a "input/ouput" type, but the hints/errors generated in both those cases would be confusing in that case, and while we could amend them
+            // for this case, it would complicate things and doesn't feel like it would feel very justified. So we merge it as an "output" type, which is the least contraining
+            // option. We do raise an hint though so users can notice this.
+            let usage = EnumTypeUsage {
+                position: EnumUsagePosition::Output,
+                examples: HashMap::new(),
+            };
+            self.error_reporter.add_hint(CompositionHint {
+                code: HintCode::UnusedEnumType.code().to_string(),
+                message: format!(
+                    "Enum type \"{}\" is defined but unused. It will be included in the supergraph with all the values appearing in any subgraph (\"as if\" it was only used as an output type).",
+                    dest.name
+                ),
+            });
+            usage
+        });
+
+        // Add all values from all sources
+        for (_, source) in sources.iter() {
+            if let Some(source) = source {
+                for value in source.values.values() {
+                    // Note that we add all the values we see as a simple way to know which values there is to consider. But some of those value may
+                    // be removed later in `merge_enum_value`
+                    if !dest.values.contains_key(&value.node.value) {
+                        dest.values.insert(value.node.value.clone(), value.clone());
+                    }
+                }
+            }
+        }
```

Comment:
This is a broader style question: This can fairly cleanly be transformed into a series of chained combinators. What are folks' preference here? This would require 4 combinators (`filter_map`, `flat_map`, `filter`, `map`) + `.extend`. Is there a cap on the length of a chain we're ok with?